File: leetcode/136.cpp
// Runtime: 16 ms
// Memory Usage: 17.2 MB

/*
  O(n) iterate through array.
*/
class Solution {
 public:
  int singleNumber(vector<int>& nums) {
    // use a sneaky property of XOR:
    // a XOR a XOR b = b
    int res = nums[0];
    for (size_t i = 1; i < nums.size(); ++i) {
      res ^= nums[i];
    }

    return res;
  }
};

File: leetcode/136.py
"""
Runtime: 128 ms, faster than 82.03% of Python3 online submissions for Single Number.
Memory Usage: 16.7 MB, less than 59.68% of Python3 online submissions for Single Number.
"""
from functools import reduce


class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        # use a hashmap: O(N) space, O(N) time
        # from collections import Counter
        # return {v:k for k,v in Counter(nums).items()}[1]

        # O(n) with no extra memory, properties of XOR
        # 1. XOR is commutative: a^b = b^a
        # 2. XOR satisfies a^a^b = b
        return reduce(lambda a, b: a ^ b, nums)
